File: ios/RNBleExtensions.swift
//
//  RNBleExtensions.swift
//  RNDsbluetooth
//
//  Created by Jaysonj on 2019/2/18.
//  Copyright © 2019 Facebook. All rights reserved.
//

import Foundation

extension DSBLEDevice {
    /// JS 对象
    ///
    /// - Returns: JS 对象
    func asJSObject() -> [String: Any] {
        return [
            "identifier": self.identifier,
            "mac": self.standardMac ?? "",
            "name": self.cbPeripheral.name ?? "",
            "rssi": self.rssi ?? -99,
            "isConnectable": true,
            "isDfu": self.state,
            "type": self.type.rawValue,
        ]
    }
    
    func debug() {
        let obj = asJSObject()
        print("Device: \(obj)")
    }
}

extension DSBLEVersion {
    /// JSON对象
    func asJSObject() -> [String: Any] {
        return [
            "version": self.version,
            "display": self.display,
            "vendor": self.vendor,
            "data": self.data
        ]
    }
}

extension DSBLEAlarm {
    
    func asJSObject() -> [String: Any] {
        return [
            "index": self.alarmNO,
            "time": self.alarmTime,
            "sunday": self.sundaySwitch,
            "monday": self.mondaySwitch,
            "tuesday": self.tuesdaySwitch,
            "wednesday": self.wednesdaySwitch,
            "thursday": self.thursdaySwitch,
            "friday": self.fridaySwitch,
            "saturday": self.saturdaySwitch,
            "switch": self.switchStatus,
        ]
    }
    
    static func asObject(dict: Dictionary<String,Any>) -> DSBLEAlarm {
        let alarm = DSBLEAlarm.init()
        alarm.alarmNO = (dict["index"] as? UInt) ?? 0
        alarm.alarmTime = (dict["time"] as? String) ?? "0800"
        alarm.switchStatus = (dict["switch"] as? Bool) ?? false
        alarm.sundaySwitch = (dict["sunday"] as? Bool) ?? false
        alarm.mondaySwitch = (dict["monday"] as? Bool) ?? false
        alarm.tuesdaySwitch = (dict["tuesday"] as? Bool) ?? false
        alarm.wednesdaySwitch = (dict["wednesday"] as? Bool) ?? false
        alarm.thursdaySwitch = (dict["thursday"] as? Bool) ?? false
        alarm.fridaySwitch = (dict["friday"] as? Bool) ?? false
        alarm.saturdaySwitch = (dict["saturday"] as? Bool) ?? false
        return alarm
    }
}

extension DSBLEAlertDistance {
    func asJSObject() -> [String: Any] {
        return [
            "farRange": self.farRange,
            "nearRange": self.nearRange,
        ]
    }
    
    static func asObject(dict: Dictionary<String,Any>) -> DSBLEAlertDistance {
        let distance = DSBLEAlertDistance.init()
        distance.farRange = (dict["farRange"] as? UInt) ?? 0
        distance.nearRange = (dict["nearRange"] as? UInt) ?? 0
        return distance
    }
}

extension DSBLEGsensor {
    func asJSObject() -> [String: Any] {
        return [
            "x": self.x,
            "y": self.y,
            "z": self.z,
        ]
    }
}

extension DSBLERestMode {
    static func asObject(dict: Dictionary<String,Any>) -> DSBLERestMode {
        let restMode = DSBLERestMode.init()
        restMode.swtich = (dict["switch"] as? Bool) ?? false
        restMode.startTime = RCTConvert.nsDate(dict["startTime"])
        restMode.endTime = RCTConvert.nsDate(dict["endTime"])
        return restMode
    }
}

extension DSBLEClimate {
    static func asObject(dict: Dictionary<String,Any>) -> DSBLEClimate {
        let climate = DSBLEClimate.init()
        climate.temperature = (dict["temperature"] as? Int) ?? 0
        climate.type =  DSBLETemperatureType.init(rawValue: (dict["type"] as? UInt) ?? 0) ?? .negative
        climate.weatherState = DSBLEWeatherState.init(rawValue: (dict["weatherState"] as? UInt) ?? 0) ?? .fine
        return climate
    }
}

extension DSBLESleepBlock {
    func asJSObject() -> [String: Any] {
        return [
            "type": self.type.rawValue,
            "value": self.value.rawValue,
            "time": self.time.timeIntervalSince1970
        ]
    }
}

extension DSBLESleepState {
    func asJSObject() -> [String: Any] {
        return [
            "state": self.state.rawValue,
            "beginTime": self.beginTime.timeIntervalSince1970,
            "endTime": self.endTime.timeIntervalSince1970,
        ]
    }
}

extension DSBLESleepInfo {
    func asJSObject() -> [String: Any] {
        var sleepState:[[String: Any]] = []
        self.sleepStates.forEach { (state) in
            sleepState.append(state.asJSObject())
        }
        return [
            "awakeDuration": self.awakeDuration,
            "beginTime": self.beginTime.timeIntervalSince1970,
            "endTime": self.endTime.timeIntervalSince1970,
            "deepSleepDuration": self.deepSleepDuration,
            "lightSleepDuration": self.lightSleepDuration,
            "sleepStates": sleepState,
        ]
    }
}

extension DSBLEStepInfo {
    func asJSObject() -> [String: Any] {
        return [
            "step": self.step,
            "type": self.type.rawValue,
            "endTime": self.time.timeIntervalSince1970,
            "startTime": self.starTtime.timeIntervalSince1970,
        ]
    }
}

extension DSBLEHeartrate {
    func asJSObject() -> [String: Any] {
        return [
            "time": self.time.timeIntervalSince1970,
            "bo": self.bo,
            "hr": self.hr,
            "type": self.type.rawValue
        ]
    }
}

extension DSBLEBloodPressure {
    func asJSObject() -> [String: Any] {
        return [
            "time": self.time.timeIntervalSince1970,
            "diastole": self.diastole,
            "systole": self.systole
        ]
    }
}

extension DSBLESport {
    func asJSObject() -> [String: Any] {
        return [
            "step": self.step,
            "sType": self.sType.rawValue,
            "time": self.time.timeIntervalSince1970,
            "type": self.type.rawValue
        ]
    }
}

extension DSBLESport2 {
    func asJSObject() -> [String: Any] {
        return [
            "cal": self.cal,
            "data": self.data,
            "endTime": self.endTime.timeIntervalSince1970,
            "startTime": self.startTime.timeIntervalSince1970,
            "sType": self.sType.rawValue,
        ]
    }
}

extension DSBLEDFUSet {
    
    static func asObejct(dict: Dictionary<String,Any>) -> DSBLEDFUSet {
        let dfuSet = DSBLEDFUSet.init()
        dfuSet.deviceMac = (dict["deviceMac"] as? String) ?? ""
        dfuSet.deviceName = (dict["deviceName"] as? String) ?? ""
        dfuSet.filePath = (dict["filePath"] as? String) ?? ""
        dfuSet.fileType = DSBLEOTAFileType.init(rawValue: (dict["fileType"] as? UInt8) ?? 0) ??  .framework
        return dfuSet
    }
}

extension DSBLESedentary {
    
    static func asObject(dict: Dictionary<String,Any>) -> DSBLESedentary {
        let sedentary = DSBLESedentary.init()
        sedentary.swtich = (dict["switch"] as? Bool) ?? false
        sedentary.sedentaryTime =  (dict["sedentaryTime"] as? UInt) ?? 0
        sedentary.startTime = RCTConvert.nsDate(dict["startTime"])
        sedentary.endTime = RCTConvert.nsDate(dict["endTime"])
        return sedentary
    }
}

extension DSBLEAppNotify {
    static func asObject(dict: Dictionary<String,Any>) -> DSBLEAppNotify {
        let notify = DSBLEAppNotify.init()
        dict.forEach { (key,value) in
            if let on = value as? Bool {
                notify.setAppNotify(key: key, value: on)
            }
        }
        return notify
    }
}

extension DSBLEWristingTime {
    static func asObject(dict: Dictionary<String,Any>) -> DSBLEWristingTime {
        let wristingTime = DSBLEWristingTime.init()
        wristingTime.startTime = RCTConvert.nsDate(dict["startTime"])
        wristingTime.endTime = RCTConvert.nsDate(dict["endTime"])
        wristingTime.wristingType = DSBLEWristingType.init(rawValue: (dict["wristingType"] as? UInt) ?? 0) ?? .off
        return wristingTime
    }
}

extension DSBLESportSet {
    static func asObject(dict: Dictionary<String,Any>) -> DSBLESportSet {
        let sportSet = DSBLESportSet.init()
        sportSet.state = DSBLESportState.init(rawValue: (dict["state"] as? UInt) ?? 0) ?? .stop
        sportSet.time = (dict["time"] as? UInt) ?? 0
        sportSet.type = DSBLESportType.init(rawValue: (dict["type"] as? UInt) ?? 0) ?? .normal
        return sportSet
    }
}

extension DSBLESportInfo {
    static func asObject(dict: Dictionary<String,Any>) -> DSBLESportInfo {
        let sportInfo = DSBLESportInfo.init()
        sportInfo.calorie = (dict["calorie"] as? UInt) ?? 0
        sportInfo.distance = (dict["distance"] as? UInt) ?? 0
        sportInfo.time = (dict["time"] as? UInt) ?? 0
        return sportInfo
    }
}

extension DSBLEBloodPressureSet {
    static func asObject(dict: Dictionary<String,Any>) -> DSBLEBloodPressureSet {
        let bpSet = DSBLEBloodPressureSet.init()
        bpSet.diastole = (dict["diastole"] as? UInt) ?? 0
        bpSet.systole = (dict["systole"] as? UInt) ?? 0
        bpSet.on = (dict["on"] as? Bool) ?? false
        return bpSet
    }
}

extension DSBLEVibrationSet {
    static func asObject(dict: Dictionary<String,Any>) -> DSBLEVibrationSet {
        let vSet = DSBLEVibrationSet.init()
        vSet.on = (dict["on"] as? Bool) ?? false
        vSet.type = DSBLEVibrationType.init(rawValue: (dict["type"] as? UInt) ?? 0) ?? .middle
        return vSet
    }
}

extension DSBLEUserInfo {
    static func asObject(dict: Dictionary<String,Any>) -> DSBLEUserInfo {
        let userinfo = DSBLEUserInfo.init()
        userinfo.age = (dict["age"] as? UInt) ?? 20
        userinfo.gender = DSBLEGender.init(rawValue: (dict["gender"] as? UInt) ?? 0) ?? .male
        userinfo.height = (dict["height"] as? UInt) ?? 170
        userinfo.weight = (dict["weight"] as? UInt) ?? 70
        return userinfo
    }
}

extension DSBLEWeather {
    static func asObject(dict: Dictionary<String,Any>) -> DSBLEWeather {
        let weather = DSBLEWeather.init()
        weather.aqi = (dict["aqi"] as? Int) ?? 170
        weather.temperature = (dict["temperature"] as? Int) ?? 170
        weather.type = DSBLETemperatureType.init(rawValue: (dict["type"] as? UInt) ?? 0) ?? .negative
        return weather
    }
}

extension DSBLEHRMonitor {
    static func asObject(dict: Dictionary<String,Any>) -> DSBLEHRMonitor {
        let hrMonitor = DSBLEHRMonitor.init()
        hrMonitor.interval = (dict["interval"] as? UInt) ?? 15
        hrMonitor.swtich = (dict["switch"] as? Bool) ?? false
        return hrMonitor
    }
}

extension DSBLEGPS {
    func asJSObject() -> Dictionary<String,Any> {
        
        return [
            "latitude": self.latitude,
            "longitude": self.longitude,
            "time": self.time.timeIntervalSince1970
        ]
    }
}

extension DSBLEPAIInfo {
    func asJSObject() -> Dictionary<String,Any> {
        return [
            "interval": self.interval,
            "time": self.time.timeIntervalSince1970,
            "type": self.type.rawValue,
            "value": self.value,
        ]
    }
}

extension DSBLEShoeStepInfo {
    func asJSObject() -> Dictionary<String,Any> {
        return [
            "duration": self.duration,
            "step": self.step,
            "type": self.type.rawValue,
            "time": self.time.timeIntervalSince1970
        ]
    }
}

extension DSBLESyncData {
    func asJSObject() -> Dictionary<String,Any> {
        var steps: [[String: Any]] = []
        self.steps.forEach { (step) in
            steps.append(step.asJSObject())
        }
        
        var hrs: [[String: Any]] = []
        self.heartratesAndBloodOxygens.forEach { (hr) in
            hrs.append(hr.asJSObject())
        }
        
        var sleeps: [[String: Any]] = []
        self.sleeps.forEach { (block) in
            sleeps.append(block.asJSObject())
        }
        
        var sports: [[String: Any]] = []
        self.sports.forEach { (sport) in
            sports.append(sport.asJSObject())
        }
        
        var sports2: [[String: Any]] = []
        self.sports2.forEach { (sport2) in
            sports2.append(sport2.asJSObject())
        }
        
        var bps: [[String: Any]] = []
        self.bloodPressures.forEach { (bp) in
            bps.append(bp.asJSObject())
        }
        
        var pais: [[String: Any]] = []
        self.pais.forEach { (pai) in
            pais.append(pai.asJSObject())
        }
        
        var gps: [[String: Any]] = []
        self.gps.forEach { (g) in
            gps.append(g.asJSObject())
        }
        
        var shoeSteps: [[String: Any]] = []
        self.shoeSteps.forEach { (shoeStep) in
            shoeSteps.append(shoeStep.asJSObject())
        }
        
        return [
            "steps" : steps,
            "heartrates": hrs,
            "sleeps": sleeps,
            "sports": sports,
            "sports2": sports2,
            "bloodPressuress": bps,
            "pais": pais,
            "gps": gps,
            "shoeSteps": shoeSteps
        ]
    }
}

File: android/src/main/java/com/jh/bluetooth/RNBleExtensions.kt
package com.jh.bluetooth

import com.desay.dsbluetooth.data.enums.*
import com.desay.dsbluetooth.data.model.*
import com.desay.dsbluetooth.manager.keep.DSBLEDevice
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.WritableMap
import java.util.Date
import kotlin.collections.HashMap

/**
 * 蓝牙库数据结构Builder
 */
object DSBLEObjectBuilder {
    fun createNewWeather(dict: HashMap<String, Any>): DSBLENewWeather {
        val temperature = (dict["temperature"] as? Int) ?: 0
        val type = if (((dict["type"] as? Int)
                        ?: 0) == 0) DSBLETemperatureType.negative else DSBLETemperatureType.positive
        val weatherState = DSBLEWeatherType.values().firstOrNull {
            it.type == ((dict["weatherState"] as? Int) ?: 0)
        }
                ?: DSBLEWeatherType.fine
        return DSBLENewWeather(type, temperature, weatherState)
    }

    fun createAlarm(dict: HashMap<String, Any>): DSBLEAlarm {
        val index = (dict["index"] as? Int) ?: 0
        val alarmTime = (dict["time"] as? String) ?: "0800"
        val switchStatus = (dict["switch"] as? Boolean) ?: false
        val sundaySwitch = (dict["sunday"] as? Boolean) ?: false
        val mondaySwitch = (dict["monday"] as? Boolean) ?: false
        val tuesdaySwitch = (dict["tuesday"] as? Boolean) ?: false
        val wednesdaySwitch = (dict["wednesday"] as? Boolean) ?: false
        val thursdaySwitch = (dict["thursday"] as? Boolean) ?: false
        val fridaySwitch = (dict["friday"] as? Boolean) ?: false
        val saturdaySwitch = (dict["saturday"] as? Boolean) ?: false
        return DSBLEAlarm(index, switchStatus, alarmTime, mondaySwitch, tuesdaySwitch,
                wednesdaySwitch, thursdaySwitch, fridaySwitch, saturdaySwitch, sundaySwitch)
    }

    fun createAlertDistance(dict: HashMap<String, Any>): DSBLEAlertDistance {
        val farRange = (dict["farRange"] as? Int) ?: 0
        val nearRange = (dict["nearRange"] as? Int) ?: 0
        return DSBLEAlertDistance(nearRange, farRange)
    }

    fun createRestMode(dict: HashMap<String, Any>): DSBLERestMode {
        val switch = (dict["switch"] as? Boolean) ?: false
        val startTime = Date(dict["startTime"] as? Long ?: 0)
        val endTime = Date(dict["endTime"] as? Long ?: 0)
        return DSBLERestMode(switch, startTime, endTime)
    }

    fun createDFUSet(dict: HashMap<String, Any>): DSBLEDFUSet {
        val deviceMac = (dict["deviceMac"] as? String) ?: ""
        val deviceName = (dict["deviceName"] as? String) ?: ""
        val filePath = (dict["filePath"] as? String) ?: ""
        val fileType = DSBLEOTAFileType.values().firstOrNull {
            it.con.toInt() == ((dict["fileType"] as? Int) ?: 0)
        } ?: DSBLEOTAFileType.Firmware
        return DSBLEDFUSet(deviceName,deviceMac,filePath,fileType)
    }

    fun createSedentary(dict: HashMap<String, Any>): DSBLESedentary {
        val switch = (dict["switch"] as? Boolean) ?: false
        val sedentaryTime = (dict["sedentaryTime"] as? Int) ?: 0
        val startTime = Date(dict["startTime"] as? Long ?: 0)
        val endTime = Date(dict["endTime"] as? Long ?: 0)
        return DSBLESedentary(sedentaryTime,switch,startTime,endTime)
    }

    fun createWristingTime(dict: HashMap<String, Any>): DSBLEWristingTime {
        val startTime = Date(dict["startTime"] as? Long ?: 0)
        val endTime = Date(dict["endTime"] as? Long ?: 0)
        val wristingType = DSBLEWristingType.values().firstOrNull { it.type == ((dict["wristingType"] as? Int)
            ?: 0)} ?: DSBLEWristingType.off
        return DSBLEWristingTime(wristingType,startTime, endTime)
    }

    fun createSportSet(dict: HashMap<String, Any>): DSBLESportSet {
        val time = (dict["time"] as? Int) ?: 0
        val state = DSBLESportState.values().firstOrNull { it.type ==  (dict["state"] as? Int) ?: 0} ?: DSBLESportState.stop
        val type = DSBLESportType.values().firstOrNull { it.type == (dict["type"] as? Int) ?: 0 } ?: DSBLESportType.Normal
        return DSBLESportSet(state, type, time)
    }

    fun createSportInfo(dict: HashMap<String, Any>): DSBLESportInfo {
        val calorie = (dict["calorie"] as? Int) ?: 0
        val distance = (dict["distance"] as? Int) ?: 0
        val time = (dict["time"] as? Int) ?: 0
        return DSBLESportInfo(time, distance, calorie)
    }

    fun createBloodPressureSet(dict: HashMap<String, Any>): DSBLEBloodPressureSet {
        val diastole = (dict["diastole"] as? Int) ?: 0
        val systole = (dict["systole"] as? Int) ?: 0
        val on = (dict["on"] as? Boolean) ?: false
        return DSBLEBloodPressureSet(on, systole, diastole)
    }

    fun createVibrationSet(dict: HashMap<String, Any>): DSBLEVibrationSet {
        val on = (dict["on"] as? Boolean) ?: false
        val type = DSBLEVibrationType.values().firstOrNull { it.type ==  (dict["type"] as? Int) ?: 0} ?: DSBLEVibrationType.middle
        return DSBLEVibrationSet(type, on)
    }

    fun createUserInfo(dict: HashMap<String, Any>): DSBLEUserInfo {
        val age = (dict["age"] as? Int) ?: 20
        val gender = DSBLEGender.values().firstOrNull { it.type == (dict["gender"] as? Int) ?: 0 } ?: DSBLEGender.male
        val height = (dict["height"] as? Int) ?: 170
        val weight = (dict["weight"] as? Int) ?: 70
        return DSBLEUserInfo(height, weight, gender, age)
    }

    fun createWeather(dict: HashMap<String, Any>): DSBLEWeather {
        val aqi = (dict["aqi"] as? Int) ?: 170
        val temperature = (dict["temperature"] as? Int) ?: 170
        val type = DSBLETemperatureType.values().firstOrNull { it.type ==  (dict["type"] as? UInt) ?: 0} ?: DSBLETemperatureType.negative
        return DSBLEWeather(aqi, type, temperature)
    }

    fun createHRMonitor(dict: HashMap<String, Any>): DSBLEHRMonitor {
        val interval = (dict["interval"] as? Int) ?: 15
        val switch = (dict["switch"] as? Boolean) ?: false
        return DSBLEHRMonitor(switch, interval)
    }

    fun createPush(dict: HashMap<String, Any>): DSBLEPushData {
        val chinese = (dict["isChinese"] as? Boolean) ?: false
        val msg = (dict["msg"] as? String)
        val time = (dict["time"] as? Int) ?: 3
        val type = DSBLEPushType.values().firstOrNull { it.type ==  (dict["type"] as? Int) ?: 0} ?: DSBLEPushType.sms
        return DSBLEPushData(chinese,msg,time,type)
    }
}

fun DSBLEDevice.asJSObject(): WritableMap {
    val result = Arguments.createMap()
    result.putString("mac", this.address)
    result.putString("name",this.device.name)
    result.putInt("rssi", this.rssi ?: -99)
    result.putBoolean("isConnectable", true)
    result.putBoolean("isDfu", this.ota)
    result.putString("type", this.type.con)
    return result
}

fun DSBLEVersion.asJSObject(): WritableMap {
    val result = Arguments.createMap()
    result.putInt("version", this.version)
    result.putString("display", this.display)
    result.putString("vendor", this.vendor)
    result.putString("data", this.data)
    return result
}

fun DSBLEAlarm.asJSObject(): WritableMap {
    val result = Arguments.createMap()
    result.putInt("index", this.alarmNO)
    result.putString("time", this.alarmTime)
    result.putBoolean("sunday", this.sundaySwitch)
    result.putBoolean("monday", this.mondaySwitch)
    result.putBoolean("tuesday", this.tuesdaySwitch)
    result.putBoolean("wednesday", this.wednesdaySwitch)
    result.putBoolean("thursday", this.thursdaySwitch)
    result.putBoolean("friday", this.fridaySwitch)
    result.putBoolean("saturday", this.saturdaySwitch)
    result.putBoolean("switch", this.switchStatus)
    return result
}

fun DSBLEAlertDistance.asJSObject(): WritableMap {
    val result = Arguments.createMap()
    result.putInt("farRange", this.farRange)
    result.putInt("nearRange", this.nearRange)
    return result
}

fun DSBLEGsensor.asJSObject(): WritableMap {
    val result = Arguments.createMap()
    result.putInt("x", this.x)
    result.putInt("y", this.y)
    result.putInt("z", this.z)
    return result
}

fun DSBLESleepBlock.asJSObject(): WritableMap {
    val result = Arguments.createMap()
    result.putInt("type", this.type.type)
    result.putInt("value", this.value.type)
    result.putInt("time", (this.time.time / 1000L).toInt())
    return result
}

fun DSBLESleepState.asJSObject(): WritableMap {
    val result = Arguments.createMap()
    result.putInt("state", this.state.type)
    result.putInt("beginTime", (this.beginTime.time / 1000L).toInt())
    result.putInt("endTime", (this.endTime.time / 1000L).toInt())
    return result
}

fun DSBLESleepInfo.asJSObject(): WritableMap {
    val result = Arguments.createMap()
    val stateArray = Arguments.createArray()
    this.sleepStates.forEach {
        stateArray.pushMap(it.asJSObject())
    }
    result.putInt("awakeDuration", this.awakeDuration)
    result.putInt("deepSleepDuration", this.deepSleepDuration)
    result.putInt("lightSleepDuration", this.lightSleepDuration)
    result.putInt("beginTime", (this.beginTime.time / 1000L).toInt())
    result.putInt("endTime", (this.endTime.time / 1000L).toInt())
    result.putArray("sleepStates", stateArray)
    return result
}

fun DSBLEStepInfo.asJSObject(): WritableMap {
    val result = Arguments.createMap()
    result.putInt("step", this.step)
    result.putInt("type", this.type.type)
    result.putInt("endTime", (this.time.time / 1000L).toInt())
    result.putInt("startTime", (this.startTime.time / 1000L).toInt())
    return result
}

fun DSBLEHeartrate.asJSObject(): WritableMap {
    val result = Arguments.createMap()
    result.putInt("hr", this.hr)
    result.putInt("bo", this.bo)
    result.putInt("type", this.type.type)
    result.putInt("time", (this.time.time / 1000L).toInt())
    return result
}

fun DSBLEBloodPressure.asJSObject(): WritableMap {
    val result = Arguments.createMap()
    result.putInt("diastole", this.diastole)
    result.putInt("systole", this.systole)
    result.putInt("time", (this.time.time / 1000L).toInt())
    return result
}

fun DSBLESport.asJSObject(): WritableMap {
    val result = Arguments.createMap()
    result.putInt("step", this.step)
    result.putInt("sType", this.sType.type)
    result.putInt("time", (this.time.time / 1000L).toInt())
    result.putInt("type", this.type.type)
    return result
}

fun DSBLESport2.asJSObject(): WritableMap {
    val result = Arguments.createMap()
    result.putInt("data", this.data)
    result.putInt("cal", this.cal)
    result.putInt("sType", this.sType.type)
    result.putInt("endTime", (this.endTime.time / 1000L).toInt())
    result.putInt("startTime", (this.startTime.time / 1000L).toInt())
    return result
}

fun DSBLEGPS.asJSObject(): WritableMap {
    val result = Arguments.createMap()
    result.putDouble("latitude",this.latitude)
    result.putDouble("longitude",this.longitude)
    result.putInt("time",(this.time.time / 1000L).toInt())
    return result
}

fun DSBLEPAIInfo.asJSObject(): WritableMap {
    val result = Arguments.createMap()
    result.putInt("interval",this.interval)
    result.putInt("type",this.type.type)
    result.putInt("time",(this.time.time / 1000L).toInt())
    result.putInt("value",this.value)
    return result
}

fun DSBLEShoeStepInfo.asJSObject(): WritableMap {
    val result = Arguments.createMap()
    result.putInt("step", this.step)
    result.putInt("type", this.type.type)
    result.putInt("time", (this.time.time / 1000L).toInt())
    result.putInt("duration", this.duration)
    return result
}

fun DSBLESyncData.asJSObject(): WritableMap {
    val result = Arguments.createMap()
    val steps = Arguments.createArray()
    this.steps.forEach { steps.pushMap(it.asJSObject()) }
    val hrs = Arguments.createArray()
    this.heartratesAndBloodOxygens.forEach { hrs.pushMap(it.asJSObject()) }
    val sleeps = Arguments.createArray()
    this.sleeps.forEach { sleeps.pushMap(it.asJSObject()) }
    val sports = Arguments.createArray()
    this.sports.forEach { sports.pushMap(it.asJSObject()) }
    val sports2 = Arguments.createArray()
    this.sports2.forEach { sports2.pushMap(it.asJSObject()) }
    val bps = Arguments.createArray()
    this.bloodPressures.forEach { bps.pushMap(it.asJSObject()) }
    val pais = Arguments.createArray()
    this.pais.forEach { pais.pushMap(it.asJSObject()) }
    val gps = Arguments.createArray()
    this.gps.forEach { gps.pushMap(it.asJSObject()) }
    val shoeSteps = Arguments.createArray()
    this.shoeSteps.forEach { shoeSteps.pushMap(it.asJSObject()) }

    result.putArray("steps",steps)
    result.putArray("heartrates",hrs)
    result.putArray("sleeps",sleeps)
    result.putArray("sports",sports)
    result.putArray("sports2",sports2)
    result.putArray("bloodPressuress",bps)
    result.putArray("pais",pais)
    result.putArray("gps",gps)
    result.putArray("shoeSteps",shoeSteps)
    return result
}










